org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/CompilationUnit.java

```diff
@@ -1577,13 +1577,13 @@ public void setOptions(Map<String, String> newOptions) {
 
 @Override
 public Map<String, String> getCustomOptions() {
-	if (this.owner != null) {
+	if (this.owner != null && DOM_BASED_OPERATIONS) {
 		try {
 			Map<String, String> customOptions = this.getCompilationUnitElementInfo().getCustomOptions();
 			IJavaProject parentProject = getJavaProject();
 			Map<String, String> parentOptions = parentProject == null ? JavaCore.getOptions() : parentProject.getOptions(true);
 			if (JavaCore.ENABLED.equals(parentOptions.get(JavaCore.COMPILER_PB_ENABLE_PREVIEW_FEATURES)) &&
-				AST.newAST(parentOptions).apiLevel() < AST.getJLSLatest()) {
+				CompilerOptions.versionToJdkLevel(parentOptions.getOrDefault(JavaCore.COMPILER_SOURCE, JavaCore.latestSupportedJavaVersion())) < CompilerOptions.versionToJdkLevel(JavaCore.latestSupportedJavaVersion())) {
```

Comment:
In case you want to keep it a bit shorter have a look at `ClassFileConstants.getLatestJDKLevel()`

With fallbacks explicit on both sides, we now see that you allow enabling preview features also for unspecified and unsupported Java versions (where the LHS will fall back to _latest_). Is that by intention? What should be the outcome?

Reply:
> Is that by intention? What should be the outcome?

Yes, but it's not a strong requirement. Basically, the goal is that for an unknown version, we failback in most cases where AST is used to the "most capable" configuration set to get an AST as helpful as possible. My interpretation is that, usually, the highest version + preview feature is the most capable set.